[DOC-845] Upgrade with rollback FAQ (#28681)
yugabyte/yugabyte-db: docs/content/preview/faq/operations-faq.md

{{<lead link="../../yugabyte-platform/manage-deployments/upgrade-software-install/">}}
For detailed information, see [Upgrade YugabyteDB](../../manage/upgrade-deployment/).
{{</lead>}}

### How long can a universe run after the upgrade but before the finalize step?

A universe can run as long as needed while the application is being validated. However, it is recommended to finalize within 3 days. Operations like flag changes are disabled during the monitoring phase.

### Are new YugabyteDB features disabled until the upgrade is finalized?

New features that change the data format on disk are disabled. Because these data-format-changing features are new, they generally aren't being used in production environments yet, so their temporary disablement has no impact on current operations. 90% of code changes are validated from the new binary itself. The aim of rollback capabilities is to catch regressions to existing features or query plans and handle them quickly (by reverting to the previous binary without any data corruption).

### Is it possible to run DDL operations during the upgrade (before finalize)?

> DDL operations are allowed for v2.20 to v2024.2 upgrades. DDL operations are only _blocked_ when the PostgreSQL version is upgraded, such as from v2024.2 (PostgreSQL 11) to v2025.1 (PostgreSQL 15).

Reviewer: ```suggestion
DDL operations are allowed during regular upgrades (v2.20 through v2024.2). DDL operations are only _blocked_ during YSQL major upgrades, where the PostgreSQL version is also upgraded, such as from v2024.2 (PostgreSQL 11) to v2025.1 (PostgreSQL 15).
```

> ### Assuming rollback is not possible post-finalize, what happens if issues are observed in the application after finalization?

Reviewer: ```suggestion
### Assuming rollback is not possible post-finalize, what do I do if I discover problems after finalization?
```

> YugabyteDB performs extensive testing to ensure there are no issues. For customers that are extremely risk averse, the recommendation is to upgrade and finalize a development environment and DR replicas (if any) first. After the data format on disk has changed, you cannot go back to old binaries—this is a technical limitation of any software that stores data to disk. The only way to rollback after finalize is to restore from a backup that was taken before the finalization (with loss of data).

Reviewer: ```suggestion
Yugabyte performs extensive testing to ensure there are no issues. After the data format on disk has changed, you cannot go back to old binaries; this is a technical limitation of any software that stores data to disk. The only way to roll back after finalize is to restore from a backup that was taken before the finalization (with loss of data).

For customers who are extremely risk averse, the recommendation is to upgrade and finalize a development/staging environment and xCluster DR replicas (if any) before upgrading your production environment.
```

> ### How does the rollback feature interact with bidirectional xCluster setups?

Reviewer: ```suggestion
### How does rollback interact with bidirectional xCluster setups?
```

> xCluster can only replicate from an old version to the new version. The target universe should be finalized before the source universe. If the source is finalized before the target, then xCluster automatically pauses itself (this only happens in certain versions that have an external data format change, such as v2024.2). For bidirectional setups, if writes are only happening on one side, then it's acceptable to upgrade and finalize the other side first and then upgrade the writing side. If both sides are taking writes, then both should be finalized at the same time; otherwise, replication in the new-to-old direction will not happen.

Reviewer: ```suggestion
xCluster can only replicate from an old version to the new version. You should finalize the target universe before the source universe. If the source is finalized before the target, then xCluster automatically pauses itself (this only happens in certain versions that have an external data format change, such as v2024.2).

For bidirectional setups, if writes are only happening on one side, then you can upgrade and finalize the other side first and then upgrade the writing side. If both sides are taking writes, then both should be finalized at the same time; otherwise, replication in the new-to-old direction will not happen.
```

{{<lead link="../../yugabyte-platform/manage-deployments/xcluster-replication/bidirectional-replication/">}}
For more information about bidirectional xCluster replication, see [Bidirectional replication using xCluster](../../yugabyte-platform/manage-deployments/xcluster-replication/bidirectional-replication/).
{{</lead>}}

### Are the xCluster bidirectional upgrade steps the same for both YSQL and YCQL?

Yes.

> ### What is the behavior of AutoFlags during the upgrade process?

Reviewer: ```suggestion
### What is the behavior of AutoFlags during an upgrade?
```

YugabyteDB Anywhere enables Volatile AutoFlags after all nodes are running the new binary version (before finalize). Only Persisted and External AutoFlags are enabled after finalization.

{{<lead link="https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/auto_flags.md">}}
For detailed information about AutoFlags, see [AutoFlags](https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/auto_flags.md).
{{</lead>}}

### If not all new features are required immediately, is there a way to selectively enable them post-finalize?

> Features have dedicated knobs to allow you to tune or disable them. For example, Cost-based Optimizer has a flag that lets you pick the mode, and tablet split has threshold flags that can be set.

Reviewer: ```suggestion
Features have dedicated flags to allow you to tune or disable them. For example, cost-based optimizer has a flag that lets you pick the mode, and tablet split has threshold flags that can be set.
```

### What exactly does the upgrade with rollback feature do?

The feature upgrades the binaries and enables new features that do not change the data format on disk. Features that do require changes to the format of data sent over the network, or stored on disk, are disabled until you finalize the upgrade.

> Because the vast majority of new changes do not modify data format, and most changes that do are from new features that are not yet in use, this temporary disabling typically has no impact on current operations. This allows you to monitor the new binary for a while (validating about 90% of code changes) to make sure there is no impact to your application and rollback to the previous version if needed.

Reviewer: ```suggestion
Because the vast majority of new changes do not modify the data format, and most changes that do are from new features that are not yet in use, this temporary disabling typically has no impact on current operations. This allows you to monitor the new version for a while (validating typically about 90% of code changes) to ensure the upgrade has not impacted your application.
```

> The aim of the rollback is to catch any regressions to existing features or query plans and handle them quickly by reverting to the previous binary without any data corruption.

Reviewer: ```suggestion
If you discover any regressions to existing features or query plans, rollback allows you to handle them quickly by reverting to the previous binary without any data corruption.
```

A universe can run as long as needed while the application is being validated. However, it is recommended to finalize within 2 days. Operations like flag changes are disabled during the monitoring phase.

### Is it possible to run DDL operations during the upgrade (before finalize)?

DDL operations are allowed during regular upgrades (major and minor). DDL operations are only _blocked_ during YSQL major upgrades, where the PostgreSQL version is also upgraded, such as from v2024.2 (PostgreSQL 11) to v2025.1 (PostgreSQL 15).

### Assuming rollback is not possible post-finalize, what do I do if I discover problems after finalization?

Yugabyte performs extensive testing to ensure there are no issues. After the data format on disk has changed, you cannot go back to old binaries; this is a technical limitation of any software that stores data to disk. The only way to roll back after finalize is to restore from a backup that was taken before the finalization (with loss of data).

For customers that are extremely risk averse, the recommendation is to upgrade and finalize a development environment and DR replicas (if any) first.

### How does rollback interact with xCluster setups?

> You can upgrade and rollback each cluster individually.

Reviewer: ```suggestion
You can upgrade and roll back each cluster individually.
```
